src/Neo/Cryptography/Base58.cs

```diff
@@ -74,19 +78,27 @@ public static byte[] Decode(string input)
             var bi = BigInteger.Zero;
             for (int i = 0; i < input.Length; i++)
             {
-                int digit = Alphabet.IndexOf(input[i]);
-                if (digit < 0)
+                if (!s_alphabetDic.Value.TryGetValue(input[i], out var digit))
                     throw new FormatException($"Invalid Base58 character '{input[i]}' at position {i}");
                 bi = bi * Alphabet.Length + digit;
             }
 
             // Encode BigInteger to byte[]
             // Leading zero bytes get encoded as leading `1` characters
-            int leadingZeroCount = input.TakeWhile(c => c == Alphabet[0]).Count();
-            var leadingZeros = new byte[leadingZeroCount];
-            if (bi.IsZero) return leadingZeros;
-            var bytesWithoutLeadingZeros = bi.ToByteArray(isUnsigned: true, isBigEndian: true);
-            return Concat(leadingZeros, bytesWithoutLeadingZeros);
+            int leadingZeroCount = LeadingBase58Zeros(input);
+            if (bi.IsZero)
+            {
+                return new byte[leadingZeroCount];
+            }
+
+            int decodedSize = bi.GetByteCount(true) + leadingZeroCount;
+
+            Span<byte> result = decodedSize <= 128
+                ? stackalloc byte[decodedSize]
+                : new byte[decodedSize];
+
+            _ = bi.TryWriteBytes(result[leadingZeroCount..], out _, true, true);
```

Comment:
Needs error checking validation. Currently this function is saying that the return value is a valid `base58` value; no matter what.